File: src-tauri/src/commands/metadata.rs
use crate::database::{ColumnInfo, IndexInfo, ForeignKeyInfo, DatabaseInfo, TableInfo, SchemaInfo, FunctionInfo, ExtensionInfo};
use crate::AppState;
use super::error::ToCommandResult;
use tauri::State;
use serde_json::Value;

#[tauri::command]
pub async fn get_databases(connection_id: String, state: State<'_, AppState>) -> Result<Vec<DatabaseInfo>, String> {
    state.connection_manager.get_databases(&connection_id).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_tables(connection_id: String, database: Option<String>, state: State<'_, AppState>) -> Result<Vec<TableInfo>, String> {
    state.connection_manager.get_tables(&connection_id, database.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_views(connection_id: String, database: Option<String>, state: State<'_, AppState>) -> Result<Vec<TableInfo>, String> {
    state.connection_manager.get_views(&connection_id, database.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_schemas(connection_id: String, database: Option<String>, state: State<'_, AppState>) -> Result<Vec<SchemaInfo>, String> {
    state.connection_manager.get_schemas(&connection_id, database.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_schema_tables(connection_id: String, database: String, schema: String, state: State<'_, AppState>) -> Result<Vec<TableInfo>, String> {
    let tables = state.connection_manager.get_tables(&connection_id, Some(&database)).await.to_cmd_result()?;
    Ok(tables.into_iter().filter(|t| t.schema.as_deref() == Some(&schema)).collect())
}

#[tauri::command]
pub async fn get_schema_views(connection_id: String, database: String, schema: String, state: State<'_, AppState>) -> Result<Vec<TableInfo>, String> {
    let views = state.connection_manager.get_views(&connection_id, Some(&database)).await.to_cmd_result()?;
    Ok(views.into_iter().filter(|v| v.schema.as_deref() == Some(&schema)).collect())
}

#[tauri::command]
pub async fn get_schema_functions(connection_id: String, database: String, schema: String, state: State<'_, AppState>) -> Result<Vec<FunctionInfo>, String> {
    tracing::info!(conn = %connection_id, db = %database, sc = %schema, "正在获取 Schema 函数...");
    state.connection_manager.get_functions(&connection_id, Some(&database), Some(&schema)).await.map_err(|e| {
        tracing::error!(err = %e, "获取 Schema 函数失败");
        e.to_string()
    })
}

#[tauri::command]
pub async fn get_schema_aggregate_functions(connection_id: String, database: String, schema: String, state: State<'_, AppState>) -> Result<Vec<FunctionInfo>, String> {
    state.connection_manager.get_aggregate_functions(&connection_id, Some(&database), Some(&schema)).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_database_extensions(connection_id: String, database: String, state: State<'_, AppState>) -> Result<Vec<ExtensionInfo>, String> {
    tracing::info!(conn = %connection_id, db = %database, "正在获取数据库扩展...");
    state.connection_manager.get_extensions(&connection_id, Some(&database)).await.map_err(|e| {
        tracing::error!(err = %e, "获取数据库扩展失败");
        e.to_string()
    })
}

#[tauri::command]
pub async fn get_table_structure(connection_id: String, table: String, database: Option<String>, schema: Option<String>, state: State<'_, AppState>) -> Result<Vec<ColumnInfo>, String> {
    state.connection_manager.get_table_structure(&connection_id, &table, schema.as_deref(), database.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_table_indexes(connection_id: String, table: String, schema: Option<String>, state: State<'_, AppState>) -> Result<Vec<IndexInfo>, String> {
    state.connection_manager.get_indexes(&connection_id, &table, schema.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_table_foreign_keys(connection_id: String, table: String, schema: Option<String>, state: State<'_, AppState>) -> Result<Vec<ForeignKeyInfo>, String> {
    state.connection_manager.get_foreign_keys(&connection_id, &table, schema.as_deref()).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_schema_indexes(connection_id: String, database: String, schema: String, state: State<'_, AppState>) -> Result<Vec<IndexInfo>, String> {
    state.connection_manager.get_schema_indexes(&connection_id, Some(&database), Some(&schema)).await.to_cmd_result()
}

#[tauri::command]
pub async fn get_create_table_ddl(connection_id: String, table: String, database: Option<String>, schema: Option<String>, state: State<'_, AppState>) -> Result<String, String> {
    state.connection_manager.get_table_ddl(&connection_id, &table, schema.as_deref(), database.as_deref()).await.to_cmd_result()
}

#[tauri::command] pub async fn get_autocomplete_data(_c: String, _d: Option<String>) -> Result<Value, String> { Ok(serde_json::json!({})) }

File: src/api/metadata.ts
import { invoke } from '@tauri-apps/api/core'
import type { 
  DatabaseInfo, TableInfo, SchemaInfo, ColumnInfo, IndexInfo, ForeignKeyInfo, FunctionInfo, ExtensionInfo 
} from '@/types/database'

export const metadataApi = {
  /**
   * 获取数据库列表
   */
  async getDatabases(connectionId: string): Promise<DatabaseInfo[]> {
    return invoke<DatabaseInfo[]>('get_databases', { connectionId })
  },

  /**
   * 获取表列表
   */
  async getTables(connectionId: string, database?: string | null): Promise<TableInfo[]> {
    return invoke<TableInfo[]>('get_tables', { connectionId, database })
  },

  /**
   * 获取视图列表
   */
  async getViews(connectionId: string, database?: string | null): Promise<TableInfo[]> {
    return invoke<TableInfo[]>('get_views', { connectionId, database })
  },

  /**
   * 获取 Schema 列表 (PostgreSQL)
   */
  async getSchemas(connectionId: string, database?: string | null): Promise<SchemaInfo[]> {
    return invoke<SchemaInfo[]>('get_schemas', { connectionId, database })
  },

  /**
   * 获取指定 Schema 下的表
   */
  async getSchemaTables(connectionId: string, database: string, schema: string): Promise<TableInfo[]> {
    return invoke<TableInfo[]>('get_schema_tables', { connectionId, database, schema })
  },

  /**
   * 获取指定 Schema 下的视图
   */
  async getSchemaViews(connectionId: string, database: string, schema: string): Promise<TableInfo[]> {
    return invoke<TableInfo[]>('get_schema_views', { connectionId, database, schema })
  },

  /**
   * 获取函数列表
   */
  async getSchemaFunctions(connectionId: string, database: string, schema: string): Promise<FunctionInfo[]> {
    return invoke<FunctionInfo[]>('get_schema_functions', { connectionId, database, schema })
  },

  /**
   * 获取聚合函数列表
   */
  async getSchemaAggregateFunctions(connectionId: string, database: string, schema: string): Promise<FunctionInfo[]> {
    return invoke<FunctionInfo[]>('get_schema_aggregate_functions', { connectionId, database, schema })
  },

  /**
   * 获取数据库扩展 (PostgreSQL)
   */
  async getDatabaseExtensions(connectionId: string, database: string): Promise<ExtensionInfo[]> {
    return invoke<ExtensionInfo[]>('get_database_extensions', { connectionId, database })
  },

  /**
   * 获取表结构
   */
  async getTableStructure(params: {
    connectionId: string,
    table: string,
    database?: string | null,
    schema?: string | null
  }): Promise<ColumnInfo[]> {
    return invoke<ColumnInfo[]>('get_table_structure', params)
  },

  /**
   * 获取表索引
   */
  async getTableIndexes(params: {
    connectionId: string,
    table: string,
    schema?: string | null
  }): Promise<IndexInfo[]> {
    return invoke<IndexInfo[]>('get_table_indexes', params)
  },

  /**
   * 获取表外键
   */
  async getTableForeignKeys(params: {
    connectionId: string,
    table: string,
    schema?: string | null
  }): Promise<ForeignKeyInfo[]> {
    return invoke<ForeignKeyInfo[]>('get_table_foreign_keys', params)
  },

  /**
   * 获取 Schema 下的所有索引
   */
  async getSchemaIndexes(connectionId: string, database: string, schema: string): Promise<IndexInfo[]> {
    return invoke<IndexInfo[]>('get_schema_indexes', { connectionId, database, schema })
  },

  /**
   * 获取创建表的 DDL
   */
  async getCreateTableDdl(params: {
    connectionId: string,
    table: string,
    database?: string | null,
    schema?: string | null
  }): Promise<string> {
    return invoke<string>('get_create_table_ddl', params)
  }
}
